### PR TITLE
Reorganize inspector layout workflow for `Control` nodes

### LINKED ISSUE #299: Refactor the SizeFlags enum of Control nodes
### Describe the project you are working on:

Any game with a UI

### Describe the problem or limitation you are having in your project:

The problem that I am having is that, either I do not understand the SizeFlags enum of Control nodes, or the enum is designed in a confusing way that allows invalid configurations.

According to the [Control node documentation](https://docs.godotengine.org/en/latest/classes/class_control.html#enumerations):

>enum **SizeFlags**:
>  * **SIZE_FILL** = **1** — Tells the parent Container to expand the bounds of this node to fill all the available space without pushing any other node. Use with size_flags_horizontal and size_flags_vertical.
>  * **SIZE_EXPAND** = **2** — Tells the parent Container to let this node take all the available space on the axis you flag. If multiple neighboring nodes are set to expand, they’ll share the space based on their stretch ratio. See size_flags_stretch_ratio. Use with size_flags_horizontal and size_flags_vertical.
>  * **SIZE_EXPAND_FILL** = **3** — Sets the node’s size flags to both fill and expand. See the 2 constants above for more information.
>  * **SIZE_SHRINK_CENTER** = **4** — Tells the parent Container to center the node in itself. It centers the control based on its bounding box, so it doesn’t work with the fill or expand size flags. Use with size_flags_horizontal and size_flags_vertical.
>  * **SIZE_SHRINK_END** = **8** — Tells the parent Container to align the node with its end, either the bottom or the right edge. It doesn’t work with the fill or expand size flags. Use with size_flags_horizontal and size_flags_vertical.

"Fill" says it "fills all the available space without pushing any other node", but this is false in testing. This sentence implies that the last item in an HBoxContainer would grow to the right, it doesn't. The actual behavior is that it fills up the reserved space.

Both "Shrink Center" and "Shrink End" say "It doesn’t work with the fill or expand size flags", and I don't think they work with each other either. Shrink Center is mutually exclusive with everything else, and Shrink End is also mutually exclusive with everything else.

I don't think there is a use case for "Expand" without "Fill", but I could be wrong. So, from what I can gather, there are only **5** valid configurations out of the 16 possible combinations:

* No size flags
* Fill only
* Fill with Expand (ExpandFill)
* Shrink Center only
* Shrink End only

Furthermore, having none of the size flags enabled basically functions as a "Shrink Begin", if I'm understanding this correctly (because it's not filling or being moved to center/end).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I think that it would make more sense to refactor this into a "dropdown"-style contiguous enum (0,1,2,3,4) instead of a "checkbox"-style bitwise enum (0,1,2,4,8), and rename some of the options. I was thinking something like this:

* Normal (same as Fill currently, and called this way because this is the default and most common)
* Expand (same as ExpandFill currently)
* Shrink Begin (same as everything unchecked currently)
* Shrink Center (same)
* Shrink End (same)

It would require a very heavy amount of refactoring which would break compatibility. For example, code that checks for Fill would need to be changed from `& SizeFlags.Fill` to `< 2` or `< SizeFlags.ShrinkBegin` or similar. Code that checks for Expand would need to be changed from `& SizeFlags.Expand` to `== SizeFlags.Expand` and code that checks for ShrinkCenter/ShrinkEnd must be `== SizeFlags.ShrinkCenter` etc.

### Describe how this feature / enhancement will help you overcome this problem or limitation:

If the SizeFlags enum was less confusing, it would help people understand Control nodes more easily.

### If this enhancement will not be used often, can it be worked around with a few lines of script?:

I think Control nodes are used very often, so making them less confusing is good.

### Is there a reason why this should be core and not an add-on in the asset library?:

SizeFlags is a core feature.

### LINKED ISSUE #2841: Redesign how user interface (Controls) scaling  is presented to the user
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

Users often complain that sizing controls (user interfaces) in Godot is complex and very confusing. Too many properties and it's not obvious what each of them does. Among the properties there is a lot of redundancy so it can take a while to correctly understand what is going on.

That said, the user interface system is extremely powerful once you understand how it works, so the goal is not so much to change it but to change how information is presented to the user to make it less confusing.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

There are currently 3 separate modes for user interface sizing and placement in Godot:

* Anchors
* Rect position and size, including rotation, scale, etc.
* Containers

image

Each serves a different purpose and use case, but all are presented at the same time to users. That said, using all of them at the same time is generally not needed:

* When _containers_ are used (control is inside a container), _anchors_ and _pos/size_ are ignored.
* When using _anchors_, _position and size_ are ignored
* Even when using containers, the size flags exposed as well as the stretch ratio vary on their availability depending on the parent container.

The idea of this proposal is to hide what is not needed, and to also allow the user to specify better the kind of workflow desired.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The first step would be to move everything position related to the top of the list, so the properties are not mixed up with other, unrelated ones.

##### Containers

The following changes will take place in the inspector:

* "Size Flags" section will be renamed to "Container Sizing"
*  In this section, the parent container will be asked for which flags are supported (not all support expand, not always in both vertical and horizontal direction, and not all support stretch ratios).
*  If the parent to a node being edited is a container, the "Positioning" (anchors/position/size/transform properties) will be hidden, as they are unused in this workflow. Only "Container Sizing" will be visible and shown on top.
* "Minimum Size" will be moved to its own section, as its useful for both operation modes (container and regular position) and will appear as second section below "Container Sizing" or "Positioning".

image

##### Positioning

All positioning related properties will be put in a larger section "Positioning". Also moved to the top. When containers are used this section goes hidden, so on top of the inspector either this or containers will be there.

_Positioning_ will have the following properties and subsections:

* *use_anchors*: Anchoring will be enabled only if this property is on. If enabled, the following sub-sections of "Positioning" will be visible:
   * Anchors
   * Margins
   Likewise, the "Rectangle" subsection (showing position and size) will be hidden if anchors are active. 
   By default, anchors will not be enabled. To use anchors, you will need to toggle them on.
   
image

* If *use_anchors* is disabled, then the "Rectangle" subsection of "Positioning" will be shown. This displays position and size.
   
image

* a "Transform" subsection of "Positioning" will be always visible (And below either "Anchoring or Rectangle"), this will contain rotation, scale, pivot, etc.
* The Clip Content property will be moved to visibility because it does not belong here. Likely renamed to "Clip to Rect"

### Is there a reason why this should be core and not an add-on in the asset library?

Not possible to do with script, should be core.

### PR DESCRIPTION
**Pull request by YuriSizov**
Closes https://github.com/godotengine/godot-proposals/issues/2841, closes https://github.com/godotengine/godot-proposals/issues/299.

Following the rationale and suggestions presented in the aforementioned proposals this provides a new, focused way to set up the layout and position property for `Control` nodes in Godot. I think it's best to present these changes in a form of a minute long overview of the new flow:

https://user-images.githubusercontent.com/11782833/142727395-2ab827d2-bee5-4ee8-8fd2-dccdbb12de44.mp4

The idea is to only suggest properties that are currently relevant when working with control nodes. Godot has two distinct positioning systems (with some permutations within): anchors and containers. Those two positioning modes are complimentary to each other, but cannot be used at the same time on the same control. This is a point of huge confusion for our users, as the editor normally doesn't guide you or prevent you from changing the wrong properties in a wrong context.

A recent PR to prevent editing properties controlled by the container helps with preventing, but doesn't really give a good cues as to why the properties are suddenly unavailable. To address that (and using the discussion from https://github.com/godotengine/godot-proposals/issues/2841) I've introduced a new common "Layout" section to the `Control` properties that encompasses everything related to positioning. It shows some universal properties by default, but also hides most of the rest beyond the selected "Layout Mode" setting.

Layout mode can have one of these 4 values:
* **Position**. This is the case for the anchors set to top-left preset and only rectangle position and size can be adjusted. This should help with the most straight-forward way to use anchors from the top-left corner of the viewport.
* **Anchors**. This allows to actually modify anchors, offsets, and grow direction directly, and can be used as a little more advanced mode on top of **Position** when you need to align nodes to different parts of the viewport.
  * Also here you can either pick a preset for anchors, offsets, and grow directions, or enter "Custom" mode and adjust them as you would before, by hand to any value you want.
* **Container**. This mode is enabled automatically for children of containers and cannot be changed. In that mode you only have size flags to configure.
  * Following https://github.com/godotengine/godot-proposals/issues/299 the size flags themselves have been reorganized to give a better idea of what combinations work and what don't. Built-in containers provide three possible states, and they are now correctly represented by the inspector: Fill/Shrinking options, Same + Expand option, No supported size flags at all.
  * This functionality is also exposed to custom containers, so you can control it with your custom container scripts (don't forget the tool mode). Note that "Shrink Begin" cannot be disabled because it's the same as having no flags. Even if your custom container doesn't pass it as an allowed flag, it will still be available.

https://user-images.githubusercontent.com/11782833/142728395-ed8e725b-86db-4a0f-bef3-17f1b2c6290e.mp4

* **Uncontrolled**. Finally, this is the mode closest to what we had before and it is enabled automatically for control nodes which are either the root node of a scene, or have a non-Control as a parent. In those cases we cannot be sure what layout mode is going to be used at runtime (especially for the roots of scenes). As such, all options are presented (anchor preset logic still applies).

-----

To help with guiding users I've added a collapsible warning to the top of the "Layout" section. In its short form it gives a quick message that can be understood and used by experienced users. If extended, it gives a short tip about what positioning properties are available.

Additionally I've introduced an "indented" rule for the inspector properties that allows to indent a group of properties with a special hint and give that indentation a little colorful side-border. I find this quite important to highlight among others those sections that become editable when the "Custom" mode is selected for anchors. I expect that this may not be a desired addition to the Inspector code, or maybe just the implementation, but I really think it brings the whole thing together on the user experience level. I've experimented a bit with different highlighting rules, or none at all, and this is really what I found to give the best visual association.

Note that every new component works well with RTL layouts as well!

-----

I've also added quick container sizing options to the toolbar. They are a bit ugly and take space, but that's something that I will hopefully solve with https://github.com/godotengine/godot-proposals/issues/3550. But something felt amiss in the toolbar without being able to quickly adjust container children as well as non-container children.

godot windows tools 64_2021-11-20_15-57-39

Those options are properly disabled following the same parent node rules as the layout options. With an explanation added via the tooltip (as it was already done before for the anchor layout presets). Speaking of which I've tried to explain better what "Keep Ratio" does with a new name and a tooltip:

godot windows tools 64_2021-11-20_15-57-52

Even with the aforementioned proposal to move to tool drawers this dropdown list is not at all practical for its own set purpose. I'm eyeing https://github.com/godotengine/godot-proposals/issues/3559 as a way to address it, albeit I would want to start with a less custom solution and just rearrange the existing icons in a rectangle, corresponding to their configuration.

-----

One technical note. The new layout mode and anchor preset inspector properties are supposed to be inspector only. I managed to remove them from the docs using `PROPERTY_USAGE_INTERNAL`, but they are still available to scripting, even though there is no purpose for them there. I don't know of a better approach to provide such inspector items, as I need an actual property to drive the rest of the properties there, and make them enabled or disabled, and I need to have it exactly where they are now. If anyone has a better suggestion for this, let me know. But this is probably not that problematic and can be left as is.

### CODE REVIEW

**Comment by golddotasksquestions** [comment]
I'm excited to play with this new UI workflow. I still hate the philosophy behind Control to bits, but this PR is looks like it could make Control nodes much more intuitive and easier to use! Thanks a lot for all the work as well as presenting it in such detail!

**Comment by KoBeWi** [comment]
When you have Anchor Mode enabled (in the toolbar) and move the control, layout mode switches to Anchors, but doesn't enable the options:
KNXDNIrPN3
Also if you undo in this state, the control won't return to its original position/size.
Also also, it would be nice if switching to Position layout mode didn't reset everything.

The expand checkbox comes without property name and doesn't have any tooltip:
image

Shouldn't the custom container sizing include all flags?
image
It's custom anyways, and here `expand` is missing. Also, while normally vertical expand is not available for HBoxContainer, you can select it from the Container toolbar:
image

It seems weird to me that setting Anchors from the toolbar sometimes doesn't have effect on control's position, but from the inspector it does:
j87vTObUAX
Especially when toolbar option will change inspector setting 🤔

> Note that "Shrink Begin" cannot be disabled because it's the same as having no flags. Even if your custom container doesn't pass it as an allowed flag, it will still be available.

It should be mentioned in the docs for `_get_allowed_flags()`. Also `h`/`v` should be `horizontal`/`vertical`, as per recent changes.

> One technical note. The new layout mode and anchor preset inspector properties are supposed to be inspector only. I managed to remove them from the docs using PROPERTY_USAGE_INTERNAL, but they are still available to scripting, even though there is no purpose for them there.

From what I see, they are stored inside meta and not available as properties. It's fine like this.
Although are they used at runtime? If they have `_edit`, I assume they have effect only in the editor, no?

**Comment by pycbouh** [comment]
> The expand checkbox comes without property name and doesn't have any tooltip:

It's all a part of the same property, vertical/horizontal size flag.

> Shouldn't the custom container sizing include all flags? It's custom anyways, and here expand is missing.

No, it only allows flags that the container supports. The reason for `Custom` to exist is to support custom containers that may work in ways that Godot built-ins don't. Custom containers now have a way to report what flags they support just like built-ins, so they can allow "Expand" if they want. We still don't want to allow invalid combinations here.

> Also, while normally vertical expand is not available for HBoxContainer, you can select it from the Container toolbar

I realized that later, but saw no point reworking the toolbar when it was already done properly with the drawers PR. I can remove "Fill & Expand" from the list, if that helps, but I've decided that until drawers are merged we can live with the same options that were previously available for containers in the Inspector.

> It seems weird to me that setting Anchors from the toolbar sometimes doesn't have effect on control's position, but from the inspector it does

IIRC I was hesitant to change the way presets from the toolbar are applied, because I'm not sure how people use all those options. I mostly don't have need for anchor mode or keeping aspect, or setting anchors without offsets. So I didn't want to break that, while still making the Inspector changes to the spec from the proposal. I'm fine making the toolbar behave the same as the Inspector. Keep in mind, that inspector also changes grow direction in anchor preset mode.

> From what I see, they are stored inside meta and not available as properties. It's fine like this.
> Although are they used at runtime? If they have `_edit`, I assume they have effect only in the editor, no?

They are supposed to be editor-only, but I can't make them fully editor-only with the flags that we have. I guess the only side-effect here would be that setting them can potentially reset some values?

**Comment by pycbouh** [comment]
For the rest: I'll take a look and fix the problems.

**Comment by reduz** [comment]
This looks very good for the most part to me. My main gripe is that I would like to re-discuss the size flags as I am not entirely sure they can be as best as they can be. Guess when you are back from vacation we can see this.

Also would appreciate a review from @groud

**Comment by madmiraal** [comment]
> Godot has two distinct positioning systems (with some permutations within): anchors and containers. Those two positioning modes are complimentary to each other, but cannot be used at the same time on the same control.

I often use anchors and offsets (margins in 3.x).
Anchors and margins

The offsets (margins) are relative to the anchor points. While the anchors provide an easy way to set the position of the control relative to the parent without using containers.

**Comment by pycbouh** [comment]
@madmiraal I'm not sure what point you are trying to convey?

**Comment by madmiraal** [comment]
> Offsets/margins are a part of the anchors system.

Sorry, my confusion. Let me read it all again and make sure I understand what this change is doing.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
I think its good and would be nice to merge and get feedback. We can discuss the size flags at another time.

File: editor/editor_inspector.cpp [hunk 1]
```diff
@@ -1216,39 +1225,78 @@ void EditorInspectorSection::_notification(int p_what) {
...
 			if (dropping && !vbox->is_visible_in_tree()) {
 				Color accent_color = get_theme_color(SNAME("accent_color"), SNAME("Editor"));
 				draw_rect(Rect2(Point2(), get_size()), accent_color, false);
 			}
+
+			// Draw section indentation.
+			if (section_indent_style.is_valid() && section_indent > 0) {
+				Rect2 indent_rect = Rect2(Vector2(), Vector2(indent_depth * section_indent_size, get_size().height));
+				if (rtl) {
+					indent_rect.position.x = get_size().width - section_indent + section_indent_style->get_margin(SIDE_RIGHT);
```
Comment (bruvzg) [inline editor/editor_inspector.cpp:1294]: Probably should be:
```suggestion
					indent_rect.position.x = get_size().width - indent_rect.size.x - section_indent_style->get_margin(SIDE_RIGHT);
```

**Review by bruvzg — APPROVED** [review]

File: editor/editor_inspector.cpp (hunk 1 above)
Comment (pycbouh) [reply editor/editor_inspector.cpp:1294]: I think it should be the left margin subtracted then to get the same result. But either way, the value should be correct, so I can change it, but only if you think it makes the logic clearer.

Comment (pycbouh) [reply editor/editor_inspector.cpp:1294]: I double-checked using a noticeably big number for a margin, and the behavior seems to be correct:

LTR
godot windows tools 64_2022-02-10_20-27-19

RTL
godot windows tools 64_2022-02-10_20-26-49

So I'll leave it as is for now.

Comment (bruvzg) [reply editor/editor_inspector.cpp:1294]: It's not really obvious without checking what `section_indent` is, but it seems to be the same, should be OK as is.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Zylann** [comment]
Problem, it seems it is now impossible to see the size of a control when it is under a container. Sure I dont want to modify it, but I want to see its width and height... like, for example, just to know how much to set the `rect_min_size` to be slightly larger than what it currently is, but got no reference...

**Comment by eh-jogos** [comment]
> @madmiraal I'm not sure what point you are trying to convey? The part that you've quoted explains that there are two systems, anchors and **containers**. Offsets/margins are a part of the anchors system.

It's a bit confusing to me because I'm not seeing how I can control the margins (or offsets) anywhere in the inspector. I don't see the "offsets" and "grow direction" groups as in the video in the beginning of the pull request description.
image

I take it that changing size and position will automatically set these "offsets" but what I want something to be "full rect" and have a "margin" of 10 on each side?
On 3.0 I could just set the margins manually, now it seems that I need to know the size of the Control container so that I can set "position" to `(10, 10)` and size to `(size_of_the_container.x - 10, size_of_the_container.y - 10)` which does not seem half as practical as just adjusting the margins.

>> It seems weird to me that setting Anchors from the toolbar sometimes doesn't have effect on control's position, but from the inspector it does

> IIRC I was hesitant to change the way presets from the toolbar are applied, because I'm not sure how people use all those options. I mostly don't have need for anchor mode or keeping aspect, or setting anchors without offsets. So I didn't want to break that, while still making the Inspector changes to the spec from the proposal. I'm fine making the toolbar behave the same as the Inspector. Keep in mind, that inspector also changes grow direction in anchor preset mode.

This also just tripped me up a lot, as I can't control the grow directions through the inspector and setting the anchors in the toolbat had no effect on it, I though the grow direction had been completely removed, until I read this comment on the PR and set the anchor through the inspector. As I'm coming from Godot 3.x my first reflex it to use the toolbar I'm already used to.

---- EDIT ----
While writing this comment I finally noticed that you have to set the anchor preset to Custom to be able to see and change "Offsets" and "Grow Direction", but "Custom" isn't available on the toolbar, so as I was only using the toolbar I never realized it.

I don't know, but I feel like there should be a better way of doing this, as while I want to be able to customize the offsets, maybe always showing offsets and grow direction on anchor mode, and if you change their value automatically set the preset to "Custom"?

Or at least setting the default preset of uncontrolled mode to custom? As what I was trying to do was just set the grow direction of some debug labels on a character to always grow up and never down like this:
image

On 3.x in this case, I don't need to care about anchors or offsets, this is pretty much a position "mode" but I still need grow direction, and it took me a while to find this pull request and find out how I could do this.

And this in fact might be a problem. If I have a control that is a child of another control, and I want to position it manually but change the grow direction, I won't be able to do it by "Position Mode", I'll have to set "Anchor Mode", then set "Custom" just to change the grow direction.

**Comment by eh-jogos** [comment]
@YuriSizov 
> @eh-jogos
> 
> > And this in fact might be a problem. If I have a control that is a child of another control, and I want to position it manually but change the grow direction, I won't be able to do it by "Position Mode", I'll have to set "Anchor Mode", then set "Custom" just to change the grow direction.
> 
> This is 100% the intention. The related proposal and this PR aimed to streamline the UI so that users are not confused by the amount of options and when and why they are relevant. Position mode is the simplest way to position a control with anchors. It's not that useful, IMO, but it'll work for people trying things without having to burden them with anchors and stuff.
> 
> The rest of your suggestions hit the same wall. Yes, you need to explicitly change the anchor mode to custom to edit those properties. That's by design, because we want to limit the options so users don't get overwhelmed and confused. If you're experienced with Godot UI, it's not a big deal to enable the custom setting. If you're not that experienced, you have less chances to break things.
> 
> There is an existing complaint that we should at least show the properties, but make them readonly. We haven't decided on that yet. But making them visible and editable, or making the defaults to be on par with 3.x would go against the whole point of this change. Yes, you will have to relearn how this part of the inspector works.

I understand the intention is to reduce clutter, and I guess that most of my suggestions do hit that wall, but I wouldn't count "grow direction" towards clutter.

It is something that works independent of anchors, can work on position mode, can work on uncontrolled mode, and being set according to the anchors is just a "quality of life" feature. 

I think it should just be a normal property, or at least visible in all cases besides "container" mode.

About the read-only suggestions, I think some cases are necessary, like I've seen somewhere about not being able to see current `rect_size` making it hard to set `rect_min_size`, or even for discoverability. 

The way things are now I don't think I would never even imagine `grow_direction` properties exist if I were a beginner, or I'd only be able to find it through the documentation of the Control node and would be wondering why the hell I can't configure it in the editor. Well, I guess improving the documentation in that regard could go a long way to helping it in that regard. 

But I still think `grow_direction` being a "sub category" of Anchor mode, and only of the "Custom" anchor preset is misplaced.

**Comment by YuriSizov** [comment]
Yes, I understand your concern. It boiled down to the idea that it's good to set Grow Direction when you set a preset, because a lot of times a certain grow direction makes a lot of sense by default for a certain preset. And in the new flow the fields that are forced by a preset are non-editable.

We can make the grow direction fields always available, change them without user's knowledge when they apply the preset and change the preset to be "Custom" if the user overrides them in turn. But the question is whether that would lead to a good experience, or if it's going to be confusing.

**Comment by eh-jogos** [comment]
I think the question is actually if it would be better/worse experience or more/less confusing in relation to the current workflow where it is hidden under the Custom anchors layout. But I guess this would need some kind of A/B testing or focused test, as I'll obviously be biased to what I think, and I don't really see grow direction as something "attached" to Anchor Presets, since it's functionality is independent on anchors or offsets.

So yeah, I in my bias I do think it would be a better experience, specially if you're using the inspector to change the Anchor Presets, and the grow direction changes at the same time when you change the Anchor Presets, as you'd be able to see that happening, it's not totally masked (I mean, at least it's not if the "grow direction" group is opened, so I understand your concern).

But I wouldn't change the Presets to Custom when you change the grow directions though. If the grow directions properties are not nested inside the anchors, as I think they shouldn't be, changing it should not affect the anchor presets at all. Again, grow direction is not part of a anchor preset, as it's functionality is not related to anchors, adjusting them together with anchor presets is just a "quality of life" thing. 

If this "quality  of life" is confusing and we need to stay consistent, I'd go for anchor presets not changing grow direction in any case at all or maybe making it optional with an editor setting like `change_grow_direction_with_anchor_presets`, or something. (Again, this would only make sense if the grow direction properties/group were not "nested" inside Anchor presets)

**Comment by golddotasksquestions** [comment]
I personally find the new Control workflow in Godot4 (as tested in Alpha9) a huuuuge improvement over the the Godot3 workflow. I find it much more intuitive and much less confusing. Thank you very much @YuriSizov!

I have to agree though with @eh-jogos on that finding the Custom mode is a bit difficult. Personally I would have expected to find it in the "Layout Mode". 
If we do really want to keep hiding it in the "Anchors Preset" (Why?), I would suggest to at least also make it available in the Toolbar dropdown as a separate item just like in the "Anchors Preset".

Most of the Control node and Control Container user flow should be reasonably intuitive now, even for complete beginners who have not used them before. A few things that are not intuitively clear to me now in Godot4 Alpha9:

`Auto Translate`: I could not figure out what this does, even once the tooltips eventually showed up. Tested with text children nodes (Layout) and without (ColorRect).

Layout direction: Locale: I suppose this could mean what type of current language I'm using? Why is this needed if we already have "Interit"? I really don't know. The documentation of the constant is not helping either: "Automatic layout direction, determined from the current locale." ... What does locale refer to or mean here in this context?

Unfortunately there is still one glaring issue I have, which is best illustrated with one of the most basic and imho most common usecase: **To center a piece of expanding text.**

Yes, I how how to do this the "Godot way", but it this is highly counter intuitive and took me years to figure out. Unfortunately it does not seem to have improved in Godot4 Alpha9 yet. In fact it got even worse. Imagine following scenario: 

You don't want to do some elaborate layouting, all you want to do is to add a bit of text (Label), make sure it's origin is centered, so if you add any amount of text it will stay centered on the origin. Sounds simple? Well in Godot3 and still in Godot4 it is not. Naturally the user will first check the Inspector, and find the "horizontal Alignment" there. however setting this property to "center" will not have the desired results, since this will just center the text relative to the Labels rect, not set the Label nodes origin to the rect center and keep it there, expanding the rect outward as desired.

So you start to learn more about the Layout options in the toolbar and obviously try "Layout: Center", now in Godot4 called "Anchor: Center". But this will result in suddenly placing the Label in the center of the screen (if the parent is a Control root node and therefore has "Full Rect") or centered on the parent origin if it is a Node2D. It will also appear centered if the parent is a Control node with a rect size of 0 (super confusing if the rect later expands). Anyway. Even though having a parent Control with 0 rect size or Node2D first looks like the solution, it becomes obvious it is not as soon as you start typing more text and the Label rect still expands only to the right. You can click the Layout/Anchor toolbar "Center" again to again center the Label rect in relation to the parent, but you would have to do this every time you change the text!

So maybe both? Go back to the Inspector "horizontal Alignment" setting and set it to center, but of course this won't help, since this still affect only how the text is aligned inside the Label rect. (a fact also easily missed by a beginner).

Instead in Godot3 you had to set the "Grow Direction" to "Both". In Godot4 Apha9, this property is now impossible hard to find. Making all of this even a greater torture: You have figure out you have to click "Anchors" in the toolbar, for some reason randomly click "keep current Ratio", and then happen to again look through all the Inspector settings and notice you have now entered the "Custom" mode, which finally reveals the "Grow Direction", which you have to know that "Both" means finally what you have thought to be a singular one button solution called something like "Center Alignment" or "Growth: Center Alignment".

This is really really bad. It's a UI flow made by the devil. 

What's worse is how this does not even work for scaling and rotation. If you now want to scale or rotate, Godot will do this from the origin which is fixed to the top left. unless you set the so called Pivot Offset. 

So every time the text changes length, even though you have figured out how to get to the "grow both" setting, you still need to be manually set the Pivot Offset using pixel values! There is no built in way to use percentages or set it to preset positions like the Anchors would do. Imagine what a nightmare this is if you want to animate rotation and scale of a centered expanding text (or anything else you want centered and the rect expanding)

[I have a feeling at the core of this convoluted mess is still the fixed origin.](https://github.com/godotengine/godot-proposals/issues/129)

Overall though, I think this is a great improvement to user friendlyness, so thanks again to @YuriSizov and everyone who reviewed and contributed!

**Comment by eh-jogos** [comment]
@golddotasksquestions 
> So you start to learn more about the Layout options in the toolbar and obviously try "Layout: Center", now in Godot4 called "Anchor: Center". But this will result in suddenly placing the Label in the center of the screen (if the parent is a Control root node and therefore has "Full Rect") or centered on the parent origin if it is a Node2D. It will also appear centered if the parent is a Control node with a rect size of 0 (super confusing if the rect later expands). Anyway. Even though having a parent Control with 0 rect size or Node2D first looks like the solution, it becomes obvious it is not as soon as you start typing more text and the Label rect still expands only to the right. You can click the Layout/Anchor toolbar "Center" again to again center the Label rect in relation to the parent, but you would have to do this every time you change the text!

Actually the way it works now, setting the Layout to center should fix it, as it will automatically set the grow direction to both on horizontal and vertical values. As @KoBeWi first pointed in this conversation and @YuriSizov confirmed, the toolbar not setting grow direction is a bug that is waiting for consensus on how to treat toolbar things, but if you set the Acnhors to Center through the inspector instead of the toolbar, it already works.

> What's worse is how this does not even work for scaling and rotation. If you now want to scale or rotate, Godot will do this from the origin which is fixed to the top left. unless you set the so called Pivot Offset.

I gather that right now, if you use a Control with `rect_size` zero as a pivot point and then put the Label inside it with Center anchors, you can just animate this "pivot" control instead of the `Label` and your problems are completely fixed. But yeah, pivot offset with access to percentage values would be good, but it's not like it's impossible right now. 

And I don't know about other game engines, but when I used flash, even though we could change the origin of stuff, sometimes the best solution was still to have a "pivot" element you can animate freely and keep your text or whatever other element properties "clean" and "default" so you could do other effects with it.

Maybe this is like the grow direction thing, and improving documentation and tutorials to showcase this should help? I know your pain and a documentation section/tutorial on "animating UI" would really help people having to figure out this things on their own. I'm reading the guidelines on how to contribute to the documentation to see if I can help with some of this.

**Comment by golddotasksquestions** [comment]
> the toolbar not setting grow direction is a bug that is waiting for consensus on how to treat toolbar things, but if you set the Acnhors to Center through the inspector instead of the toolbar, it already works.

I see. Well, I can confirm the bug. 
While this does slightly improve things, it still does not fix the issue of this being a terrible workflow for something as stupidly simple and common place as trying to center growing text or other content. You would still need to know you have to have a parent with a rect size of 0 ... From a user perspective: Why is this necessary?

And you would also have to know you can rotate and scale the node you want to rotate and scale, but only the parent, which again, has to meet certain criteria to ever work as a pivot.

This is really backwards stupid and the opposite of a intuitive beginner friendly design.

I honestly doubt though there is much one can do about this without completely rethinking Control nodes at their very core. So my hope for Godot4 is to at least make these processes as transparent and automated as possible.

>And I don't know about other game engines, but when I used flash, even though we could change the origin of stuff, sometimes the best solution was still to have a "pivot" element you can animate freely and keep your text or whatever other element properties "clean" and "default" so you could do other effects with it.

Yes extra pivots "in between" nodes are cool very useful and have their place. But adding extra nodes should not be a necessity for the most basic tasks like centering, rotation and scaling. Godot already has origin an pivot built into every Control. It's just the user workflow implementation is completely shit in Godot 3 and prior and has not been changed since.

**Comment by eh-jogos** [comment]
@golddotasksquestions 
> I see. Well, I can confirm the bug. While this does slightly improve things (as soon as it works), it still does not fix the underlying issue of this being a terrible workflow for something as stupidly simple and common place as trying to center growing text or other content. You would still need to know you have to have a parent with a rect size of 0 ... From a user perspective: Why is this necessary? Why is this not a one button action?

you don't need to add another node as a pivot **just** to get grow directions working, you can just go to the inspector, set the Anchor preset to Center or if you want to have more control, set it to Custom and change the grow direction properties

I agree this is not super beginner friendly which was what I was talking about grow direction not being tied to anchors as it is useful and works independent of anchors, but it is also not as hard as you say. The Label can be a child of any other node and just to center the text this will always work, though currently it needs to be done in the inspector instead of the toolbar.

To animate the text, or UI in general is another story. And a tutorial/documentation with these common tips could go a long way in easing this pains even if not the ideal solution.

But I still agree that percentage based pivot offset would help on some cases.

**Comment by golddotasksquestions** [comment]
> you don't need to add another node as a pivot **just** to get grow directions working,

I did not claim you would.

What I did say is you need another parent node with 0 rect size in order to center your node (to continously grow, scale and rotate from center). 
And yes you can work around this by setting the `pivot_offset` to halve the rect size. But imho this is even more stupid, counter intuitive, complicated and hard to learn and teach, because it is a completely opaque, backwards back and forth design principle as I tried to explain [here](https://github.com/godotengine/godot-proposals/issues/129#issuecomment-539652979) and [here](https://github.com/godotengine/godot-proposals/issues/129#issuecomment-539671793). 
This has been an issue in Godot 3 and it still an issue in Godot4 Alpha 9.

Like I said, I don't assume this will be fixed with Godot4, but I really hope we at least can make this process a lot more transparent and automated (one button click upfront and and center, instead of 5 clicks into cascading menus who's purpose and discoverability is totally opaque for beginners)

**Pull request merged**